infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/SinglePortHostnameBuilder.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.docker.server.mapping;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.StringJoiner;
+
+/**
+ * Produces host names in form:
+ * server-<serverName>.<machineName>.<workspaceId>.<external_or_internal_address>.nip.io
+ *
+ * @author Max Shaposhnik (mshaposh@redhat.com)
+ */
+public class SinglePortHostnameBuilder {
+
+  private final String externalAddress;
+  private final String internalAddress;
+
+  public SinglePortHostnameBuilder(String externalAddress, String internalAddress) {
+    this.externalAddress = externalAddress;
+    this.internalAddress = internalAddress;
+  }
+
+  /**
+   * Constructs hostname from given params.
+   *
+   * @param serverName optional server name
+   * @param machineName optional machine name
+   * @param workspaceID optional workspace ID
+   * @return composite hostname
+   */
+  public String build(String serverName, String machineName, String workspaceID) {
+    StringJoiner joiner = new StringJoiner(".");
+    if (serverName != null) {
+      joiner.add("server-" + serverName.replace('/', '-'));
```

Comment:
What if multiple dashes happen, is it valid host that works in single port case?